'PEP440 and "version always up to date" tests' after 'Simplified Version and added support for PEP440'

ceball opened the pull request.

ceball committed:
Rough check for warning about invalid version.

ceball committed:
More detailed checks for version.

ceball committed:
Use right python on travis.

ceball committed:
Test develop install.

ceball committed:
Fix travis syntax.

ceball committed:
Added comment.

ceball committed:
Fixed path error.

ceball committed:
Minor: change numbering scheme to make external description easier.

ceball committed:
Insert some git resets before independent tests.

ceball committed:
Travis syntax fixup

ceball committed:
Updated desired version scheme.

ceball committed:
Just test py36 for now.

ceball committed:
Travis syntax?

ceball committed:
Ignore appveyor for now.

ceball committed:
Travis syntax?

ceball commented on scripts/tmpverify:
```diff
@@ -32,14 +32,15 @@ else:
+predicted_version = predicted_version.strip() # Strip off trailing newline
 
-assert autover.__version__==predicted_version, '%s != %s'%(autover.__version__,predicted_version)
+assert autover.__version__.startswith(predicted_version), '%r does not start with %r' % (autover.__version__,predicted_version)
```
Could you explain this change?

jlstevens committed:
Updated and fixed get_setup_version functions

jlstevens committed:
Updated tmpverify script

jlstevens commented:
@ceball The test build is now passing but the 'Conda_package' build is failing with:

```
File "/home/travis/build/ioam/autover/conda.recipe/meta.yaml", line 3, in top-level template code
    version: {{ environ['GIT_DESCRIBE_TAG'].lstrip('v') }}.{{ GIT_BUILD_STR }}
TypeError: 'NoneType' object is not callable
```

Which should have nothing to do with autover, but something to do with the meta.yaml definition/templating.

I made three basic changes:

1. Made sure ``__version__`` is always a string (solution discussed in  #14)
2. Fixed handling of file paths - the ``tmpverify`` script works inside a different current working directory. I switch to using ``__file__`` which fixes this (this was a painful one to debug!)
3. Offered different policies of what to do about a dirty repo in setup.py: ``'report'``, ``'strip'``, ``'raise'``. 

  * The ``'report'`` option is the old policy that reports if dirty. 
  *The ``'strip'`` policy never reports '-dirty' (what I used in this PR and why I updated tmpverify to use startswith). 
  * Finally, you can call ``get_setup_version`` with ``dirty='raise'`` if you want to block things from happening when the repo is dirty (e.g when building packages). Obviously you will need to only do this sometimes (e.g for ``setup.py sdist``) and needs autover as a build dependency.

The three policy options should offer the necessary flexibility to handle dirty repositories in setup.py.

ceball committed:
Get conda package version (and some other metadata) from setup.py.

jlstevens commented:
Yay! It went green!

ceball committed:
Untested test to start testing some scenarios.

ceball committed:
Merge branch 'pep440_fix' into pep440-after-pep440_fix

ceball committed:
Fix travis syntax.

ceball committed:
Fix previous commit.

ceball committed:
Fix git tagging.

ceball committed:
Switch to v prefix for tag.

ceball committed:
More test scenarios.

ceball committed:
Minor clarifications; should be no behavior change.

ceball committed:
Added another test scenario.

ceball committed:
Added commented-out start of test for package that depends on autover.

ceball committed:
Merge branch 'pep440_fix' into pep440-after-pep440_fix

ceball committed:
Update expected version to test tags.

ceball committed:
Simplify now there's no predicted version going on.

ceball committed:
Named the tests. Uncommented pkg_depends test (although unfinished)

ceball committed:
Clarified test names.

ceball committed:
Fixed test name error.

ceball committed:
Fix travis org.

ceball committed:
Removed duplicate travis entry.

ceball committed:
Commented out problematic travis entry.

ceball committed:
Added test for release tag plus commits.

ceball committed:
Increase clone depth to get a tag.

ceball committed:
Fixed typo.

ceball committed:
Merge branch 'pep440_fix' into pep440-after-pep440_fix

ceball committed:
Switch back to equals test rather than startswith.

ceball committed:
Get some info for debugging.

ceball committed:
Fixed mixed up test names.

ceball committed:
No longer need to str(__version__).

ceball committed:
Merge branch 'pep440_fix' into pep440-after-pep440_fix

ceball committed:
Resume creation of testing various scenarios.

ceball committed:
Travis debugging.

ceball committed:
Found the travis bug.

ceball committed:
Allow tmpverify to work for modules other than autover.

ceball committed:
More work on pkg_depends example.

ceball committed:
Fixed running of tmpverify now module name required.

ceball committed:
Allow example pkg_depends to find installed autover.

ceball committed:
Added pkg_bundle example, plus fixes for pkg_depends.

ceball committed:
Add example pkg_params, plus fixes.

ceball committed:
Travis experiment.

ceball committed:
Undo previous commit.

ceball committed:
Minor simplification, should be no change to behavior.

ceball committed:
minor simplification using a script.

ceball committed:
Add some commits to test.

ceball committed:
Fixed script name.

ceball committed:
Fixed error introduced by simplification script a few commits back. Removed presumably no longer needed git resets. Added test for pip+git for depends package.

ceball committed:
Test more combinations.

ceball committed:
Restore appveyor.

ceball committed:
Merge branch 'pep440_fix' into pep440-after-pep440_fix

ceball committed:
Try out bundling autover.

ceball committed:
Removed unused import. And just use setuptools.

ceball committed:
Try again to bundle...

ceball committed:
Updated pkg_depend example.

ceball committed:
Build autover package and use that as depdency for pkg_depend.

ceball committed:
Removed duplicate section [ci skip].

ceball committed:
Autodownload didn't seem to work.

ceball committed:
Fix travis build dir.

ceball committed:
Merge branch 'pep440_fix' into pep440-after-pep440_fix

ceball committed:
Increase depth again.

ceball committed:
Added git describe to help with debugging.

jlstevens committed:
Using more specific glob match for tags

ceball committed:
Merge branch 'pep440_fix' into pep440-after-pep440_fix

ceball committed:
Disable tests of getting via param for now.

jlstevens committed:
Added fix for non-existent remote

ceball committed:
Re-enable auto download test.

jlstevens committed:
Removed other fallbacks from pkg_depend

jlstevens committed:
Added debug output

ceball committed:
.version was being missed from package.

ceball committed:
Don't use symlinks, just copy the script into examples.

ceball committed:
Removed outdated comment [ci skip].

ceball committed:
Bundle version with pkg_bundle rather than copying in (so git+pip will work). Also add scripts/tmpverify to packages for same reason.

ceball committed:
Building dependency and passing it through for tox is unnecessary (tox does that itself).

ceball committed:
Revert "Building dependency and passing it through for tox is unnecessary (tox does that itself)."

This reverts commit f62226293cc83095960ee18d9ff6e95b4f3ab7d2.

Turns out I was wrong :(

ceball committed:
Added comments.

ceball committed:
Added some debug prints.

ceball committed:
More debugging.

ceball committed:
More debug prints

ceball committed:
More debugging.

ceball committed:
Minor simplification; should be no change to behavior.

ceball committed:
Clarified when test is skipped.

ceball committed:
Something is confusing about the paths. Eventually this all needs cleaning up.

ceball committed:
Updated example setup.pys to match autover.

ceball committed:
Update branch name.

jlstevens committed:
Fixed imports in test __init__.py files

jlstevens committed:
Including .version in pkg_bundle

ceball committed:
Fix directory test runs in.

jlstevens committed:
Fixed incorrect package_data

jlstevens commented:
Yay! All the travis tests are passing at last!

jlstevens committed:
Stripped out debugging print statements from version

jlstevens commented:
Just to make sure we don't forget:

1. ``ref`` of the embed function will need to be updated to an appropriate git tag.
2. This is needed in setup.py ``    package_data = {'pkg_bundle': ['.version']},`` or alternatively the ``MANIFEST`` (is it either/or...or are both needed?)
3. There is a fallback import in the package ``__init__.py`` files and the JSON fallback needs to be readded.
4. We should switch on Python 2 tests in the build matrix.

jlstevens committed:
Removed unnecessary warning

jlstevens committed:
Merge branch 'pep440_fix' into pep440-after-pep440_fix

ceball committed:
Initial pass at using dodo, to be able to test locally and on windows.

ceball committed:
Removed mistaken extra commands for testing autover itself.

ceball committed:
Fix travis temp dir.

ceball committed:
Missed .gitignore files from previous commit.

ceball committed:
Removed unused example.

ceball committed:
Fix travis temp dir.

ceball committed:
Fixed mistake in previous commit.

ceball committed:
Use doit to simplify conda package creation.

ceball committed:
Added test of pkg_bundle conda package (no autover dependency).

ceball committed:
Fixed typos.

ceball committed:
Added test of pkg_depend conda package (autover dependency at build and run time).

Also added missing info to setup.py for pkg_depend and pkg_bundle.

ceball committed:
Added some comments.

ceball committed:
Added python 2.7 example tests.

ceball committed:
Run python 2.7 tests of autover itself.

ceball committed:
Try just testing autover itself with python 2.7.

ceball committed:
Try to fix running of py27 tests.

ceball committed:
Travis simplification.

ceball committed:
Fixed pyflakes/name errors.

ceball committed:
Lint checks.

ceball committed:
Switch tmpverify from plain script to console script, for win support.

ceball committed:
pyflakes fixes

ceball committed:
Fix running tmpverify after change to script type.

ceball committed:
Fixed script location.

ceball committed:
Removed 'tmpverify' tests of autover itself.

Assuming unit tests have already/will be added to autover where necessary:

  1. autover gives a version that works for python's packaging (i.e. pep 440 compliant string)
  2. post-release versioning is working such that v0.2.0-5-g85da374 is transformed to 0.2.0.post5+g85da374
  3. only need to specify the version via the tag, and setup.py + __init__.py are up to date
  4. (works for release tag, release tag + commits, prerelease tag, prerelease tag + commits)

In any case, the example package tests cover the above too.

ceball committed:
Removed leftover install command.

ceball committed:
Removed tmpverify from autover conda package.

ceball committed:
Fix running of nosetests for conda package.

jlstevens commented:
@ceball I think I would like to merge this PR once python 2 is added to the build matrix and all the tests are green. Then we can tidy things up in other, much smaller PRs. Does that make sense to you?

ceball commented:
No problem, you are free to merge this any time.

jlstevens commented:
Great! Very happy to have these tests. Now we just need to enable the python 2 tests for everything and maybe add more unit tests. We could enable mac tests on travis but I see that as optional...

ceball commented:
I think enabling mac tests would be a bad idea for this project. Unlike windows and linux tests, mac tests will take a long time to start up and run. Meanwhile, the extra coverage provided by running on mac seems like it would be pretty small for this (dependency free) project, unless I'm missing something?
